src/assign.rs

```diff
@@ -58,25 +54,456 @@ pub struct Assignment<'a> {
+            _ => assign_scalar(token, ptr, buf, value, src)?,
+        };
+        match assigned {
+            Assigned::Done(assignment) => {
+                return Ok(assignment);
+            }
+            Assigned::Continue {
+                next_buf,
+                next_value,
+                same_src,
+            } => {
+                buf = next_buf;
+                src = same_src;
+                value = next_value;
+                ptr = tail;
+            }
+        }
+        offset += 1 + tok_len;
+    }
+
+    // Pointer is root, we can replace `dest` directly
+    let replaced = Some(core::mem::replace(value, src.into()));
+    Ok(Assignment {
+        assigned: value,
+        replaced,
+        assigned_to: buf,
+    })
+}
+
+/// Indicates error occurred during an assignment
+#[derive(Debug, PartialEq, Eq)]
+pub enum AssignError {
+    /// A `Token` within the `Pointer` failed to be parsed as an array index.
+    FailedToParseIndex {
+        /// Offset of the partial pointer starting with the invalid index.
+        offset: usize,
+        /// The source [`ParseIndexError`]
+        source: ParseIndexError,
+    },
+    /// An `Index` `Token` within the `Pointer` was out of bounds of the
+    /// target array.
+    OutOfBounds {
+        /// Offset of the partial pointer starting with the invalid index.
+        offset: usize,
+        /// The source [`OutOfBoundsError`]
+        source: OutOfBoundsError,
+    },
+}
+impl fmt::Display for AssignError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::FailedToParseIndex { offset, .. } => {
+                write!(f, "failed to parse index at offset {}", offset)
+            }
+            Self::OutOfBounds { offset, .. } => {
+                write!(f, "index at offset {} out of bounds", offset)
+            }
+        }
+    }
+}
+#[cfg(feature = "std")]
+impl std::error::Error for AssignError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        match self {
+            Self::FailedToParseIndex { source, .. } => Some(source),
+            Self::OutOfBounds { source, .. } => Some(source),
+        }
+    }
+}
+
+enum Assigned<'v> {
+    Done(Assignment<'v>),
+    Continue {
+        next_buf: PointerBuf,
+        next_value: &'v mut Value,
+        same_src: Value,
+    },
+}
+
+fn assign_array<'v>(
+    token: Token<'_>,
+    remaining: &Pointer,
+    mut buf: PointerBuf,
+    array: &'v mut Vec<Value>,
+    src: Value,
+    offset: usize,
+) -> Result<Assigned<'v>, AssignError> {
+    // parsing the index
+    let idx = token
+        .to_index()
+        .map_err(|source| AssignError::FailedToParseIndex { offset, source })?
+        .for_len_incl(array.len())
+        .map_err(|source| AssignError::OutOfBounds { offset, source })?;
+    buf.push_back(idx.into());
+
+    debug_assert!(idx <= array.len());
+
+    if idx < array.len() {
```

Comment:
Could also take advantage of the fact that `Index` is an enum and branch here on the variant.

Reply:
this is so `assign_value` can continue the traversal.